src/azul/service/manifest_service.py

```diff
@@ -2005,3 +2009,94 @@ def qualify(qualifier, column_name, index=None):
             bundle_tsv_writer.writerow(row)
+
+
+class VerbatimManifestGenerator(FileBasedManifestGenerator):
+
+    @property
+    def content_type(self) -> str:
+        return 'application/jsonl'
+
+    @classmethod
+    def file_name_extension(cls) -> str:
+        return 'jsonl'
+
+    @classmethod
+    def format(cls) -> ManifestFormat:
+        return ManifestFormat.verbatim_jsonl
+
+    @property
+    def entity_type(self) -> str:
+        return 'files'
+
+    @property
+    def included_fields(self) -> list[FieldPath]:
+        # This is only used when searching the aggregates, which are only used
+        # to perform a "join" on the replicas index. Therefore, we only need the
+        # "keys" used for the join.
+        return [
+            ('entity_id',),
+            ('contents', self.implicit_hub_type, 'document_id')
+        ]
+
+    @property
+    def implicit_hub_type(self) -> str:
+        return self.service.metadata_plugin(self.catalog).implicit_hub_type
+
+    def _replica_keys(self) -> Iterable[dict[str, str]]:
+        request = self._create_request()
+        for hit in request.scan():
+            yield {
+                # Most replicas track their hubs explicitly, however...
+                'hub_ids': hit['entity_id'],
+                # ... for projects and datasets, there are too many hubs to
+                # track them all in the replica, so they are instead retrieved
+                # by entity ID.
+                'entity_id': one(one(hit['contents'][self.implicit_hub_type])['document_id'])
+            }
+
+    def _all_replicas(self) -> Iterable[JSON]:
+        emitted_replica_ids = set()
+        page_size = 100
+        for page in chunked(self._replica_keys(), page_size):
+            for replica in self._join_replicas(page):
+                # A single replica may have many hubs. To prevent replicas from
+                # being emitted more than once, we need to keep track of
+                # replicas already emitted.
+                replica_id = replica.meta.id
+                if replica_id not in emitted_replica_ids:
+                    yield replica.contents.to_dict()
+                    # Note that this will be zero for replicas that use implicit
+                    # hubs, in which case there are actually many hubs
+                    explicit_hub_count = len(replica.hub_ids)
+                    # We don't have to track the IDs of replicas with only one
+                    # hub, since we know that there are no other hubs that could
+                    # cause their re-emission.
+                    if explicit_hub_count != 1:
+                        emitted_replica_ids.add(replica_id)
+
+    def _join_replicas(self,
+                       keys_page: Iterable[dict[str, str]]
+                       ) -> Iterable[Hit]:
+        terms_by_field = defaultdict(set)
+        for keys in keys_page:
+            for field, term in keys.items():
```

Comment:
I find the tight coupling with _replica_keys somewhat opaque. You know there are two entries in each dict and what they are.  Yet looping over them hides the static nature of that dictionary and makes the code harder to follow. Someone looking at _replica_keys may not understand why the dict is structured that way until they see this code here. It's not too bad because they are in close proximity, but I would still want to try an attrs instance with `.hub_id` and `.implicit_hub_id` (not sure about the names).